// @aws-sdk/client-pricing/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var schema = require('@smithy/core/schema');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "pricing",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class PricingClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultPricingHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class PricingServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, PricingServiceException.prototype);
    }
}

class AccessDeniedException extends PricingServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
        this.Message = opts.Message;
    }
}
class ExpiredNextTokenException extends PricingServiceException {
    name = "ExpiredNextTokenException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ExpiredNextTokenException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ExpiredNextTokenException.prototype);
        this.Message = opts.Message;
    }
}
class InternalErrorException extends PricingServiceException {
    name = "InternalErrorException";
    $fault = "server";
    $retryable = {};
    Message;
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalErrorException.prototype);
        this.Message = opts.Message;
    }
}
class InvalidNextTokenException extends PricingServiceException {
    name = "InvalidNextTokenException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "InvalidNextTokenException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidNextTokenException.prototype);
        this.Message = opts.Message;
    }
}
class InvalidParameterException extends PricingServiceException {
    name = "InvalidParameterException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
        this.Message = opts.Message;
    }
}
class NotFoundException extends PricingServiceException {
    name = "NotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "NotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, NotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class ThrottlingException extends PricingServiceException {
    name = "ThrottlingException";
    $fault = "client";
    $retryable = {
        throttling: true,
    };
    Message;
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.Message = opts.Message;
    }
}
class ResourceNotFoundException extends PricingServiceException {
    name = "ResourceNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
        this.Message = opts.Message;
    }
}

const _ADE = "AccessDeniedException";
const _AN = "AttributeName";
const _ANt = "AttributeNames";
const _AV = "AttributeValue";
const _AVL = "AttributeValueList";
const _AVt = "AttributeValues";
const _CC = "CurrencyCode";
const _DS = "DescribeServices";
const _DSR = "DescribeServicesRequest";
const _DSRe = "DescribeServicesResponse";
const _ED = "EffectiveDate";
const _ENTE = "ExpiredNextTokenException";
const _F = "Filter";
const _FF = "FileFormat";
const _FFi = "FileFormats";
const _FV = "FormatVersion";
const _Fi = "Field";
const _Fil = "Filters";
const _GAV = "GetAttributeValues";
const _GAVR = "GetAttributeValuesRequest";
const _GAVRe = "GetAttributeValuesResponse";
const _GP = "GetProducts";
const _GPLFU = "GetPriceListFileUrl";
const _GPLFUR = "GetPriceListFileUrlRequest";
const _GPLFURe = "GetPriceListFileUrlResponse";
const _GPR = "GetProductsRequest";
const _GPRe = "GetProductsResponse";
const _IEE = "InternalErrorException";
const _INTE = "InvalidNextTokenException";
const _IPE = "InvalidParameterException";
const _LPL = "ListPriceLists";
const _LPLR = "ListPriceListsRequest";
const _LPLRi = "ListPriceListsResponse";
const _M = "Message";
const _MR = "MaxResults";
const _NFE = "NotFoundException";
const _NT = "NextToken";
const _PL = "PriceList";
const _PLA = "PriceListArn";
const _PLJI = "PriceListJsonItems";
const _PLr = "PriceLists";
const _RC = "RegionCode";
const _RNFE = "ResourceNotFoundException";
const _S = "Services";
const _SC = "ServiceCode";
const _SJPLJI = "SynthesizedJsonPriceListJsonItem";
const _SL = "ServiceList";
const _Se = "Service";
const _T = "Type";
const _TE = "ThrottlingException";
const _U = "Url";
const _V = "Value";
const _a = "application/json";
const _c = "client";
const _e = "error";
const _hE = "httpError";
const _mT = "mediaType";
const _s = "server";
const _sm = "smithy.ts.sdk.synthetic.com.amazonaws.pricing";
const n0 = "com.amazonaws.pricing";
var SynthesizedJsonPriceListJsonItem = [0, n0, _SJPLJI, { [_mT]: _a }, 0];
var AccessDeniedException$ = [-3, n0, _ADE,
    { [_e]: _c, [_hE]: 401 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(AccessDeniedException$, AccessDeniedException);
var AttributeValue$ = [3, n0, _AV,
    0,
    [_V],
    [0]
];
var DescribeServicesRequest$ = [3, n0, _DSR,
    0,
    [_SC, _FV, _NT, _MR],
    [0, 0, 0, 1]
];
var DescribeServicesResponse$ = [3, n0, _DSRe,
    0,
    [_S, _FV, _NT],
    [() => ServiceList, 0, 0]
];
var ExpiredNextTokenException$ = [-3, n0, _ENTE,
    { [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ExpiredNextTokenException$, ExpiredNextTokenException);
var Filter$ = [3, n0, _F,
    0,
    [_T, _Fi, _V],
    [0, 0, 0], 3
];
var GetAttributeValuesRequest$ = [3, n0, _GAVR,
    0,
    [_SC, _AN, _NT, _MR],
    [0, 0, 0, 1], 2
];
var GetAttributeValuesResponse$ = [3, n0, _GAVRe,
    0,
    [_AVt, _NT],
    [() => AttributeValueList, 0]
];
var GetPriceListFileUrlRequest$ = [3, n0, _GPLFUR,
    0,
    [_PLA, _FF],
    [0, 0], 2
];
var GetPriceListFileUrlResponse$ = [3, n0, _GPLFURe,
    0,
    [_U],
    [0]
];
var GetProductsRequest$ = [3, n0, _GPR,
    0,
    [_SC, _Fil, _FV, _NT, _MR],
    [0, () => Filters, 0, 0, 1], 1
];
var GetProductsResponse$ = [3, n0, _GPRe,
    0,
    [_FV, _PL, _NT],
    [0, [() => PriceListJsonItems, 0], 0]
];
var InternalErrorException$ = [-3, n0, _IEE,
    { [_e]: _s, [_hE]: 500 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(InternalErrorException$, InternalErrorException);
var InvalidNextTokenException$ = [-3, n0, _INTE,
    { [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(InvalidNextTokenException$, InvalidNextTokenException);
var InvalidParameterException$ = [-3, n0, _IPE,
    { [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(InvalidParameterException$, InvalidParameterException);
var ListPriceListsRequest$ = [3, n0, _LPLR,
    0,
    [_SC, _ED, _CC, _RC, _NT, _MR],
    [0, 4, 0, 0, 0, 1], 3
];
var ListPriceListsResponse$ = [3, n0, _LPLRi,
    0,
    [_PLr, _NT],
    [() => PriceLists, 0]
];
var NotFoundException$ = [-3, n0, _NFE,
    { [_e]: _c, [_hE]: 400 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(NotFoundException$, NotFoundException);
var PriceList$ = [3, n0, _PL,
    0,
    [_PLA, _RC, _CC, _FFi],
    [0, 0, 0, 64 | 0]
];
var ResourceNotFoundException$ = [-3, n0, _RNFE,
    { [_e]: _c, [_hE]: 404 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ResourceNotFoundException$, ResourceNotFoundException);
var Service$ = [3, n0, _Se,
    0,
    [_SC, _ANt],
    [0, 64 | 0], 1
];
var ThrottlingException$ = [-3, n0, _TE,
    { [_e]: _c, [_hE]: 429 },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ThrottlingException$, ThrottlingException);
var PricingServiceException$ = [-3, _sm, "PricingServiceException", 0, [], []];
schema.TypeRegistry.for(_sm).registerError(PricingServiceException$, PricingServiceException);
var AttributeValueList = [1, n0, _AVL,
    0, () => AttributeValue$
];
var Filters = [1, n0, _Fil,
    0, () => Filter$
];
var PriceListJsonItems = [1, n0, _PLJI,
    0, [() => SynthesizedJsonPriceListJsonItem,
        0]
];
var PriceLists = [1, n0, _PLr,
    0, () => PriceList$
];
var ServiceList = [1, n0, _SL,
    0, () => Service$
];
var DescribeServices$ = [9, n0, _DS,
    0, () => DescribeServicesRequest$, () => DescribeServicesResponse$
];
var GetAttributeValues$ = [9, n0, _GAV,
    0, () => GetAttributeValuesRequest$, () => GetAttributeValuesResponse$
];
var GetPriceListFileUrl$ = [9, n0, _GPLFU,
    0, () => GetPriceListFileUrlRequest$, () => GetPriceListFileUrlResponse$
];
var GetProducts$ = [9, n0, _GP,
    0, () => GetProductsRequest$, () => GetProductsResponse$
];
var ListPriceLists$ = [9, n0, _LPL,
    0, () => ListPriceListsRequest$, () => ListPriceListsResponse$
];

class DescribeServicesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSPriceListService", "DescribeServices", {})
    .n("PricingClient", "DescribeServicesCommand")
    .sc(DescribeServices$)
    .build() {
}

class GetAttributeValuesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSPriceListService", "GetAttributeValues", {})
    .n("PricingClient", "GetAttributeValuesCommand")
    .sc(GetAttributeValues$)
    .build() {
}

class GetPriceListFileUrlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSPriceListService", "GetPriceListFileUrl", {})
    .n("PricingClient", "GetPriceListFileUrlCommand")
    .sc(GetPriceListFileUrl$)
    .build() {
}

class GetProductsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSPriceListService", "GetProducts", {})
    .n("PricingClient", "GetProductsCommand")
    .sc(GetProducts$)
    .build() {
}

class ListPriceListsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSPriceListService", "ListPriceLists", {})
    .n("PricingClient", "ListPriceListsCommand")
    .sc(ListPriceLists$)
    .build() {
}

const paginateDescribeServices = core.createPaginator(PricingClient, DescribeServicesCommand, "NextToken", "NextToken", "MaxResults");

const paginateGetAttributeValues = core.createPaginator(PricingClient, GetAttributeValuesCommand, "NextToken", "NextToken", "MaxResults");

const paginateGetProducts = core.createPaginator(PricingClient, GetProductsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListPriceLists = core.createPaginator(PricingClient, ListPriceListsCommand, "NextToken", "NextToken", "MaxResults");

const commands = {
    DescribeServicesCommand,
    GetAttributeValuesCommand,
    GetPriceListFileUrlCommand,
    GetProductsCommand,
    ListPriceListsCommand,
};
const paginators = {
    paginateDescribeServices,
    paginateGetAttributeValues,
    paginateGetProducts,
    paginateListPriceLists,
};
class Pricing extends PricingClient {
}
smithyClient.createAggregatedClient(commands, Pricing, { paginators });

const FilterType = {
    ANY_OF: "ANY_OF",
    CONTAINS: "CONTAINS",
    EQUALS: "EQUALS",
    NONE_OF: "NONE_OF",
    TERM_MATCH: "TERM_MATCH",
};

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.AccessDeniedException = AccessDeniedException;
exports.AccessDeniedException$ = AccessDeniedException$;
exports.AttributeValue$ = AttributeValue$;
exports.DescribeServices$ = DescribeServices$;
exports.DescribeServicesCommand = DescribeServicesCommand;
exports.DescribeServicesRequest$ = DescribeServicesRequest$;
exports.DescribeServicesResponse$ = DescribeServicesResponse$;
exports.ExpiredNextTokenException = ExpiredNextTokenException;
exports.ExpiredNextTokenException$ = ExpiredNextTokenException$;
exports.Filter$ = Filter$;
exports.FilterType = FilterType;
exports.GetAttributeValues$ = GetAttributeValues$;
exports.GetAttributeValuesCommand = GetAttributeValuesCommand;
exports.GetAttributeValuesRequest$ = GetAttributeValuesRequest$;
exports.GetAttributeValuesResponse$ = GetAttributeValuesResponse$;
exports.GetPriceListFileUrl$ = GetPriceListFileUrl$;
exports.GetPriceListFileUrlCommand = GetPriceListFileUrlCommand;
exports.GetPriceListFileUrlRequest$ = GetPriceListFileUrlRequest$;
exports.GetPriceListFileUrlResponse$ = GetPriceListFileUrlResponse$;
exports.GetProducts$ = GetProducts$;
exports.GetProductsCommand = GetProductsCommand;
exports.GetProductsRequest$ = GetProductsRequest$;
exports.GetProductsResponse$ = GetProductsResponse$;
exports.InternalErrorException = InternalErrorException;
exports.InternalErrorException$ = InternalErrorException$;
exports.InvalidNextTokenException = InvalidNextTokenException;
exports.InvalidNextTokenException$ = InvalidNextTokenException$;
exports.InvalidParameterException = InvalidParameterException;
exports.InvalidParameterException$ = InvalidParameterException$;
exports.ListPriceLists$ = ListPriceLists$;
exports.ListPriceListsCommand = ListPriceListsCommand;
exports.ListPriceListsRequest$ = ListPriceListsRequest$;
exports.ListPriceListsResponse$ = ListPriceListsResponse$;
exports.NotFoundException = NotFoundException;
exports.NotFoundException$ = NotFoundException$;
exports.PriceList$ = PriceList$;
exports.Pricing = Pricing;
exports.PricingClient = PricingClient;
exports.PricingServiceException = PricingServiceException;
exports.PricingServiceException$ = PricingServiceException$;
exports.ResourceNotFoundException = ResourceNotFoundException;
exports.ResourceNotFoundException$ = ResourceNotFoundException$;
exports.Service$ = Service$;
exports.ThrottlingException = ThrottlingException;
exports.ThrottlingException$ = ThrottlingException$;
exports.paginateDescribeServices = paginateDescribeServices;
exports.paginateGetAttributeValues = paginateGetAttributeValues;
exports.paginateGetProducts = paginateGetProducts;
exports.paginateListPriceLists = paginateListPriceLists;
